### DISCUSSION TITLE
How would you recommend implementing and upstreaming an assembly parser?

### DISCUSSION THREAD

**Discussion by ColinKennedy** [nvim-treesitter/nvim-treesitter#5485] [General]
I'm thinking of writing an assembly parser and if I did write it I thought maybe to upstream it to nvim-treesitter since my first parser was added pretty easily (thanks to the maintainers for that, btw).

But assembly comes in several different flavors. There's AT&T Intel, and ARM of course but even lesser known Motorola Assembly which has its own syntax.

Would an assembly parser have to be fully generic for all of these flavors? Surely not, right? If so, how would someone recommend upstreaming it to Neovim? Would it end up being added as "asm_intel" for one flavor and "asm_foo" for others? Some thoughts on this would be appreciated.

**Comment by clason** [comment]
Specifically for nvim-treesitter, I would look at current filetype detection; it's easy to assign a parser to a filetype, but if there's no dedicated filetype, it's very difficult.

**Reply by ColinKennedy** [reply]
I'm not sure if there's a dedicated file type for the different assembly syntaxes. I've only ever seen .asm as an extension and I've seen AT&T + Intel syntax used for either. Though there might actually be "recommended" extensions to differentiate them that I'm unaware of.

I know extension != file type in Vim but since that's mainly how Vim decides what to auto-set, it's tricky.

**Reply by clason** [reply]
Mainly but not exclusively. There are also checks that look at contents.

But I don't know much about assembler; in principle, it's perfectly fine to add a parser without filetype association and leave it to users (or dedicated plugins) to `vim.treesitter.start(bufnr, parser)` manually.

Regarding name: whatever makes it easiest for people to find them, so I would go by their actual names.

**Comment by ColinKennedy** [comment]
Update: I found a good resource on this topic: https://github.com/github-linguist/linguist/blob/7ca3799b8b5f1acde1dd7a8dfb7ae849d3dfb4cd/lib/linguist/languages.yml#L407

Though the list above may not be 100% convention it's probably a good estimation about which extensions developers use for which assembly languages.

I went through all of the assembly file types and noted these conflicts:

h2. File Extension Conflicts:
"Assembly" and "Motorola 68K Assembly"
- .asm
- .i
- .inc

"Unix Assembly" and "Motorola 68K Assembly"
- .s

Luckily there's less overlap than I'd thought. Depending on how it's done, it may even be possible to have 1 parser that handles more than one convention. Then the minimal conflicts wouldn't be an issue in any case since the parser can read the files either way.

Here's the full list of assembly extensions if anyone's curious. Note that duplicates indicate "there was more than one parser that subscribed to this file extension" (a conflict)

<details>
<summary>All extensions</summary>

```
- ".a51"
- ".agc"
- ".asm"
- ".asm"
- ".c++-objdump"
- ".c++objdump"
- ".c-objdump"
- ".cpp-objdump"
- ".cppobjdump"
- ".cxx-objdump"
- ".d-objdump"
- ".i"
- ".i"
- ".inc"
- ".inc"
- ".ms"
- ".nas"
- ".nasm"
- ".objdump"
- ".pasm"
- ".s"
- ".s"
- ".wast"
- ".wat"
- ".x68"
```
</details>

**Comment by xasc** [comment]
My recommendation would be to look at the GCC file type conventions ([here](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Overall-Options.html#Options-Controlling-the-Kind-of-Output)).

This adds `file.S` and `file.sx` to the list. Both represent files that contain both assembler instructions and C preprocessor directives. On the other hand, `file.i` is typically reserved for preprocessed (intermediate) C/C++ source files.
